File: shiftreduce.cpp

#include<bits/stdc++.h>
using namespace std;
int z = 0, i = 0, j = 0, c = 0;
char a[16], ac[20], stk[15], act[10];
void check()
{
	strcpy(ac,"REDUCE TO E -> ");
	for(z = 0; z < c; z++)
	{
		if(stk[z] == 's')
		{
			printf("%ss", ac);
			stk[z] = 'E';
			stk[z + 1] = '\0';
			printf("\n$%s\t%s$\t", stk, a);
		}
	}

	for(z = 0; z < c; z++)
	{
		if(stk[z] == 'a')
		{
			printf("%sa", ac);
			stk[z] = 'E';
			stk[z + 1] = '\0';
			printf("\n$%s\t%s$\t", stk, a);
		}
	}
	for(z = 0; z < c - 2; z++)
	{
		if(stk[z] == 'E' && stk[z + 1] == '+' &&
								stk[z + 2] == 'E')
		{
			printf("%sE+E", ac);
			stk[z]='E';
			stk[z + 1]='\0';
			stk[z + 1]='\0';
			printf("\n$%s\t%s$\t", stk, a);
			i = i - 2;
		}
	}
		for(z = 0; z < c - 2; z++)
	{
		if(stk[z] == 'E' && stk[z + 1] == '-' &&
								stk[z + 2] == 'E')
		{
			printf("%sE-E", ac);
			stk[z]='E';
			stk[z + 1]='\0';
			stk[z + 1]='\0';
			printf("\n$%s\t%s$\t", stk, a);
			i = i - 2;
		}
	}
			for(z = 0; z < c - 2; z++)
	{
		if(stk[z] == 'E' && stk[z + 1] == '/' &&
								stk[z + 2] == 'E')
		{
			printf("%sE/E", ac);
			stk[z]='E';
			stk[z + 1]='\0';
			stk[z + 1]='\0';
			printf("\n$%s\t%s$\t", stk, a);
			i = i - 2;
		}
	}
	return ;
}
int main()
{
	printf("GRAMMAR is -\nE->E/E \nE->E-E \nE->E+E \nE->s \nE->a");
	strcpy(a,"s/s-a-s/s");
	c=strlen(a);
	strcpy(act,"SHIFT");
	printf("\nstack \t input \t action");
	printf("\n$\t%s$\t", a);
	for(i = 0; j < c; i++, j++)
	{
		printf("%s", act);
		stk[i] = a[j];
		stk[i + 1] = '\0';
		a[j]=' ';
		printf("\n$%s\t%s$\t", stk, a);
		check();
	}
	check();
	if(stk[0] == 'E' && stk[1] == '\0')
		printf("Accept\n");
	else
		printf("Reject\n");
}

File: shift_reduce.py
gram = {
	"S":["IF b THEN s ELSE a","(IF b THEN a"]
}
starting_terminal = "S"
inp = "IF b THEN (IF b THEN b) ELSE b$"

stack = "$"
print(f'{"Stack": <40}'+"|"+f'{"Input Buffer": <30}'+"|"+f'Parsing Action')
print(f'{"-":-<100}')

while True:
	action = True
	i = 0
	while i<len(gram[starting_terminal]):
		if gram[starting_terminal][i] in stack:
			stack = stack.replace(gram[starting_terminal][i],starting_terminal)
			print(f'{stack: <40}'+"|"+f'{inp: <30}'+"|"+f'Reduce S->{gram[starting_terminal][i]}')
			i=-1
			action = False
		i+=1
	if len(inp)>1:
		stack+=inp[0]
		inp=inp[1:]
		print(f'{stack: <40}'+"|"+f'{inp: <30}'+"|"+f'Shift')
		action = False
	if inp == "$" and stack == ("$"+starting_terminal):
		print(f'{stack: <40}'+"|"+f'{inp: <30}'+"|"+f'Accepted')
		break

	if action:
		print(f'{stack: <40}'+"|"+f'{inp: <30}'+"|"+f'Rejected')
		break
